corehq/project_limits/gauge.py

```diff
@@ -0,0 +1,188 @@
+        """
+        if not self.gauge_definition:
+            # no throttling if config to throttle is not set.
+            return True
+        computed_value = None
+        if self.gauge_definition.actionable_metric == AVG:
+            computed_value = self.gauge.avg()
+        else:
+            computed_value = self.gauge.max()
+        return computed_value < self.gauge_definition.acceptable_value
+
+    @property
+    def gauge_definition(self):
+        """
+        Should return an instace of subclass of ThrottleDefinition
+        """
+        return self.gauge_definition_fn(self.gauge.feature_key)
+
+
+class PillowLagGaugeLimiter(GaugeLimiter):
+
+    def _has_pillow_reported_recently(self):
+        time_window = 15  # time window in minutes in which throttling should be applied
+        time_before_window = datetime.now(tz=timezone.utc) - timedelta(minutes=time_window)
+        return self.gauge.get_last_reported_time() > time_before_window
+
+    def allow_usage(self):
+        if not self.gauge_definition:
+            # No throttling if config to throttle is not set
+            return True
+        return not (self._is_ideal_throttle_condition() and self._has_pillow_reported_recently())
+
+    def _is_ideal_throttle_condition(self):
+        """
+        The ideal throttle conditions are -
+        - If throttler is enabled in the config AND
+        - If we have reported values from scopes AND
+        - The follwing conditions are considered -
+            - If both max_value and average_value are set on the Defintion model,
+            then both should be breached in order to throttle.
+            - If only max_value is defined, then only max_value would be observed.
+            - If only average_value is set, then only average value would be observed.
+        """
+        if not self.gauge_definition.is_enabled:
+            # if throttling is disabled, don't throttle
+            return False
+
+        max_observed_value = self.gauge.max()
+        avg_observed_value = self.gauge.avg()
+
+        if max_observed_value is None or avg_observed_value is None:
+            # No throttling if we don't have any values
+            return False
+
+        max_value = self.gauge_definition.max_value
+        average_value = self.gauge_definition.average_value
+
+        if max_value is not None and average_value is not None:
+            # If the max and average values are both set, consider both before throttling
+
+            if max_observed_value > max_value and avg_observed_value > average_value:
+                return True
+        elif max_value is not None:
+            # If only max value is set, throttle based on max value of gauge.
+            if max_observed_value > max_value:
+                return True
+        elif average_value is not None:
+            # If only averable value is set, throttle based on average value gauge.
+            if avg_observed_value > average_value:
+                return True
```

Comment:
nit: I would clean this up since once in a certain condition, you can return whether true or false.
```suggestion
        if max_value is not None and average_value is not None:
            return max_observed_value > max_value and avg_observed_value > average_value
        elif max_value is not None:
            return max_observed_value > max_value
        elif average_value is not None:
            return avg_observed_value > average_value
```

Reply:
Yeah. This looks much better. Thanks for the suggestion.